lib/src/view/analysis/tree_view.dart

```diff
@@ -162,124 +131,696 @@ class _InlineTreeViewState extends ConsumerState<AnalysisTreeView> {
+
+  /// Callbacks for when the user interacts with the tree view, e.g. selecting a different move.
+  AnalysisController notifier,
+});
+
+/// Sidelines are usually rendered on a new line and indented.
+/// However sidelines are rendered inline (in parantheses) if the side line has no branching and is less than 6 moves deep.
+bool _displaySideLineAsInline(ViewBranch node, [int depth = 0]) {
+  if (depth == 6) return false;
+  if (node.children.isEmpty) return true;
+  if (node.children.length > 1) return false;
+  return _displaySideLineAsInline(node.children.first, depth + 1);
+}
+
+/// Returns whether this node has a sideline that should not be displayed inline.
+bool _hasNonInlineSideLine(ViewNode node) =>
+    node.children.length > 2 ||
+    (node.children.length == 2 && !_displaySideLineAsInline(node.children[1]));
+
+/// Splits the mainline into parts, where each part is a sequence of moves that are displayed on the same line.
+/// A part ends when a mainline node has a sideline that should not be displayed inline.
+Iterable<List<ViewNode>> _mainlineParts(ViewRoot root) =>
+    [root, ...root.mainline]
+        .splitAfter(_hasNonInlineSideLine)
+        .takeWhile((nodes) => nodes.firstOrNull?.children.isNotEmpty == true);
+
+/// Displays a tree-like view of a PGN game's moves.
+///
+/// For example, the PGN 1. e4 e5 (1... d5) (1... Nc6) 2. Nf3 Nc6 (2... a5) 3. Bc4 * will be displayed as:
+/// 1. e4 e5                      // [_MainLinePart]
+/// |- 1... d5                    // [_SideLinePart]
+/// |- 1... Nc6                   // [_SideLinePart]
+/// 2. Nf3 Nc6 (2... a5) 3. Bc4   // [_MainLinePart], with inline sideline
+/// Short sidelines without any branching are displayed inline with their parent line.
+/// Longer sidelines are displayed on a new line and indented.
+/// The mainline is split into parts whenever a move has a non-inline sideline, this corresponds to the [_MainLinePart] widget.
+/// Similarly, a [_SideLinePart] contains the moves sequence of a sideline where each node has only one child.
+class _PgnTreeView extends StatefulWidget {
+  const _PgnTreeView({
+    required this.root,
+    required this.rootComments,
+    required this.params,
+  });
+
+  /// Root of the PGN tree
+  final ViewRoot root;
+
+  /// Comments associated with the root node
+  final IList<PgnComment>? rootComments;
+
+  final _PgnTreeViewParams params;
+
+  @override
+  State<_PgnTreeView> createState() => _PgnTreeViewState();
+}
+
+typedef _Subtree = ({
+  _MainLinePart mainLinePart,
+
+  /// This is nullable since the very last mainline part might not have any sidelines.
+  _IndentedSideLines? sidelines,
+  bool containsCurrentMove,
+});
+
+class _PgnTreeViewState extends State<_PgnTreeView> {
+  /// Caches the result of [_mainlineParts] to avoid recalculating it on every rebuild.
+  List<List<ViewNode>> mainlineParts = [];
+
+  /// Caches the top-level subtrees, where each subtree is a [_MainLinePart] and its sidelines.
```

Comment:
We could add a precision here: this cache is a `Widget` cache obtained from the framework `build` method. The framework will skip `build` of each subtree where the widget reference is the same, this is the real perf optimisation.

Since the other cache above (`mainlineParts`) is not a widget cache it would be good to add this precision to avoid confusion.